subsys/bluetooth/host/direction.c

```diff
@@ -7,15 +7,113 @@
 #include <assert.h>
 
 #include <bluetooth/hci.h>
+#include <bluetooth/l2cap.h>
 #include <bluetooth/conn.h>
+#include <bluetooth/direction.h>
 #include <sys/byteorder.h>
 
+#include "hci_core.h"
 #include "conn_internal.h"
+#include "direction_internal.h"
 
 #define BT_DBG_ENABLED IS_ENABLED(CONFIG_BT_DEBUG_HCI_CORE)
 #define LOG_MODULE_NAME bt_df
 #include "common/log.h"
 
+/* @brief Antenna information for LE Direction Finding */
+struct bt_le_df_ant_info {
+	/* Bitfiled holding optional switching and sampling rates */
+	uint8_t switch_sample_rates;
+	/* available antennae number */
+	uint8_t num_ant;
+	/* Maximum supported antennae switching pattern length */
+	uint8_t max_switch_pattern_len;
+	/* Maximum length of CTE in 8[us] units */
+	uint8_t max_cte_len;
+};
+
+static struct bt_le_df_ant_info df_ant_info;
+
+#define DF_SUPP_TEST(feat, n)                   ((feat) & BIT((n)))
+
+#define DF_AOD_TX_1US_SUPPORT(supp)             (DF_SUPP_TEST(supp, \
+						BT_HCI_LE_1US_AOD_TX))
+#define DF_AOD_RX_1US_SUPPORT(supp)             (DF_SUPP_TEST(supp, \
+						BT_HCI_LE_1US_AOD_RX))
+#define DF_AOA_RX_1US_SUPPORT(supp)             (DF_SUPP_TEST(supp, \
+						BT_HCI_LE_1US_AOA_RX))
+
+/* @brief Function sets CTE parameters for connectionless transmission
+ *
+ * @param[in] params    Pointer to CTE parameters object
+ *
+ * @return Zero in case of success, other value in case of failure.
+ */
+static int hci_df_set_cl_cte_tx_params(const struct bt_le_ext_adv *adv,
+				const struct bt_le_df_adv_cte_tx_params *params)
+{
+	struct bt_hci_cp_le_set_cl_cte_tx_params *cp;
+	struct net_buf *buf;
+
+	/* If AoD is not enabled, ant_ids are ignored by controller:
+	 * BT Core spec 5.2 Vol 4, Part E sec. 7.8.80.
+	 */
+	if (params->cte_type == BT_HCI_LE_AOD_CTE_1US ||
+	    params->cte_type == BT_HCI_LE_AOD_CTE_2US) {
+
+		if (!BT_FEAT_LE_ANT_SWITCH_TX_AOD(bt_dev.le.features)) {
+			return -EINVAL;
+		}
+
+		if (params->cte_type == BT_HCI_LE_AOD_CTE_1US &&
+		    !DF_AOD_TX_1US_SUPPORT(df_ant_info.switch_sample_rates)) {
+			return -EINVAL;
+		}
+
+		if (params->num_ant_ids < BT_HCI_LE_SWITCH_PATTERN_LEN_MIN ||
+		    params->num_ant_ids > BT_HCI_LE_SWITCH_PATTERN_LEN_MAX ||
+		    !params->ant_ids) {
+			return -EINVAL;
+		}
+		__ASSERT_NO_MSG((sizeof(*cp) + params->num_ant_ids) <
```

Comment:
Instead of asserting we could check the valid range.
0x02 to 0x4B. that should cover the overflow.

Reply:
Valid range is actually verified few lines above. Because of that assert was redundant and removed.